practice/rain.py

```diff
@@ -2,129 +2,135 @@
+    date_amount_tuple_int = date_amount_tuple[0], int(date_amount_tuple[1])
+    return date_amount_tuple_int
+
+def create_dict_from_list_of_dates_and_rain(y):
     """Creates a dict from the list.
 
-    >>> dict_dates([['test', 1]])
-    {'test': 1}
+    >>> create_dict_from_list_of_dates_and_rain([('12-MAY-1999', 3)])
+    {'12-MAY-1999': 3}
     """
     d = {key: value for (key, value) in y}
-    d = {str(k): int(v) for k, v in d.items()}
     return d
 
 
-def most_rain(d):
+def most_rain(p):
     """Returns the key for max of dict daily totals.
 
-    >>> most_rain({'test': 1})
-    'test'
+    >>> most_rain([('10-DEC-1988', 1), ('14-JAN-1999', 10)])
+    '14-JAN-1999'
     """
-    day_with_most_rain = max(d, key=d.get)
+    dict_of_parsed_lines = create_dict_from_list_of_dates_and_rain(p)
+    day_with_most_rain = max(dict_of_parsed_lines, key=dict_of_parsed_lines.get)
     return day_with_most_rain
 
 
-def create_year_list(temp_list2):
+def get_year_from_date_amount(date_amount):
+    """
+    >>> get_year_from_date_amount(('10-DEC-1988', 1))
+    '1988'
+    """
+    year_key = date_amount[0][-4:]
+    return year_key
+
+
+def get_year_with_most_rain(parsed_lines):
+    """Returns year with max amount of rain.
+
+    >>> get_year_with_most_rain([('10-DEC-1988', 1), ('14-JAN-1999', 10), ('01-OCT-1988', 4), ('05-MAY-1999', 3)])
+    '1999'
+    """
+    total_by_year_list = create_year_list(parsed_lines)
+    total_by_year_dict = create_year_dict(total_by_year_list)
+
+    total_by_year_max = max(total_by_year_dict, key=total_by_year_dict.get)
+    return total_by_year_max
+
+
+def create_year_list(parsed_lines):
     """Slices the list of dates into years
 
-    >>> create_year_list([['Dec1999', 'Party over! OOPS! out of time']])
+    >>> create_year_list([['DEC-1999', 'Party over! OOPS! out of time']])
     [['1999', 'Party over! OOPS! out of time']]
     """
-    temp_list3 = []
-    for date, rain in temp_list2:
+    list_of_years_and_rainfall = []
+    for date, rain in parsed_lines:
         date_sliced = date[-4:]
-        temp_list3 += [[date_sliced, rain]]
-    return temp_list3
+        list_of_years_and_rainfall += [[date_sliced, rain]]
+    return list_of_years_and_rainfall
 
 
-def create_year_dict(temp_list3):
+def create_year_dict(list_of_year_and_rainfall):
     """Creates a dict from year list and sums rain value.
 
     >>> create_year_dict([['1999', 1], ['1999', 2]])
     {'1999': [3]}
     """
     yearly_totals = {}
-    for date_sliced, rain in temp_list3:
+    for date_sliced, rain in list_of_year_and_rainfall:
         if date_sliced not in yearly_totals:
             yearly_totals[date_sliced] = []
         yearly_totals[date_sliced] += [rain]
-        total_by_year = dict(zip(yearly_totals.keys(), [[sum(item)] for item in yearly_totals.values()]))
+    total_by_year = dict(zip(yearly_totals.keys(), [[sum(item)] for item in yearly_totals.values()]))
     return total_by_year
```

Comment:
Yes, I know it seems like more steps, but the resulting code is shorter because you don't have to re-implement the grouping logic. Mainly I want you to practice using `group_by()` since having the ability to think in those abstractions is a key part of the class.